### ISSUE TITLE
No more than 30,000 operation can't be performed in one transaction

### ISSUE THREAD

**Issue by Zuchos** [scredis/scredis#45]
Thre is a problem in `Protocol` class. When a transaction is executed (in `Protoco.send`) locks on `concurrentOpt` are aquired. The issue is when the number of operations in transaction + 1 is greater than the `MaxConcurrentRequestsOpt` that is defaulted to 30,000 the operation hangs forever, since that's the way the `semaphore.acquire(count)` works.
Of course the limit might be adjusted, but when the limit is reached, `Protocol` will **freeze the whole execution**.

Solution would be to perform check if the number of requested concurrent operations reached the limit value.

Besides that I wonder why every single operation in transaction is treated as concurrent operation? The transaction is executed in a single thread.

**Comment by kpbochenek** [comment]
@Zuchos There are 2 issues raised here. I will refer to the first one and try investigate more a second one.

PR  #48 should fix mentioned issue. In detail for:
- inTransaction -> returned future will be failed with RedisTansactionBuilderException[IllegalArgumentException]
- withTransaction -> exception will be thrown RedisTansactionBuilderException[IllegalArgumentException]

I will let this PR be open till tomorrow and if you have any concerns about it please comment there, if not I will merge it and release next version.

**Comment by kpbochenek** [comment]
Released `2.1.9` with mentioned fix, still not investigated second issue.

**Comment by Zuchos** [comment]
This problem hit me again so I made investigation on my own.
1. The variable 
```
 # Maximum number of overall concurrently processing requests. This parameter directly
    # influences the memory consumption of scredis. Reducing it too much will impact performance
    # negatively.
    max-concurrent-requests = 30000
```
is not concurrency fuse, its rather protecting against memory consumption.
2. The issue with `reacquiring` to many locks on semaphore persists. If 2 transaction will try to acquire lock concurrently, they will pass the if condition and then one or both will fail on acquiring them. It should be replaced with `tryAcquire` with timeout.
3. We decided to turn off that semaphore and monitor limit memory consumption on the host.

**Comment by kpbochenek** [comment]
I don't see how blocking infinitely can manifest here. 
First you do check `semaphoreAccessOption.exists(_.permitsLimit < count)` that you can't have more actions in transaction than let's say 30k, if not simply exception is thrown.
For transaction you do: `acquire(1 + transaction.requests.size)`
Which with max limit 30k and 2 transactions 20k you will jump into situation where one transaction will be executing and other one will be waiting on a semaphore, but as soon as  first transaction ends second one is executed.
Internally we do atomic lock `    semaphoreAccessOption.foreach(_.semaphore.acquire(count))` for a number of transactions not one by one, there is no option to deadlock(I think :) )
I can imagine something can freeze for a moment if there are lots of big transactions but I would expect everything to eventually be executed.
Could you provide direct example of transactions and their processings by scredis that could cause it to hang forever?

**Comment by kpbochenek** [comment]
I am going to close this issue, if you have steps to reproduce this problem please open a new issue

**Comment by Zuchos** [comment]
Sorry for not responding so long. Yeah, you are right, it will not block infinitely, but one transaction will block execution of another for a while and there is no control of how long this call will be blocked. Using `tryAcquire` with timeout would give control over it.

My point was that the name of variable `max-concurrent-requests` and its purpose are different. AFAIU all commands in transaction are executed in one Future so there is no concurrency between them. Transaction will take longer than single request and take more memory to store the list of commands and their responses, but it has nothing to do with concurrency access to the resource (redis). So here single concurrent command is treated the same way as a command in the batch the transaction. 

Maybe it's all cause by the ambiguity of the naming convention (transaction, has number of requests, which are not real requests, rather commands in one batch request).
